hoodie-utilities/src/main/java/com/uber/hoodie/utilities/deltastreamer/HoodieDeltaStreamer.java

```diff
@@ -163,16 +175,38 @@ public void sync() throws Exception {
     }
     log.info("Checkpoint to resume from : " + resumeCheckpointStr);
 
-    // Pull the data from the source & prepare the write
-    Pair<Optional<JavaRDD<GenericRecord>>, String> dataAndCheckpoint = source.fetchNewData(
-        resumeCheckpointStr, cfg.sourceLimit);
+    final JavaRDD<GenericRecord> avroRDD;
+    final String checkpointStr;
+    if (transformer != null) {
+      // Transformation is needed. Fetch New rows in Row Format, apply transformation and then convert them
+      // to generic records for writing
+      Pair<Optional<Dataset<Row>>, String> dataAndCheckpoint = source.fetchNewDataInRowFormat(
+          resumeCheckpointStr, cfg.sourceLimit);
+
+      if ((!dataAndCheckpoint.getLeft().isPresent()) || (dataAndCheckpoint.getLeft().get().count() == 0)) {
+        log.info("No new data, nothing to commit.. ");
+        return;
+      }
 
-    if (!dataAndCheckpoint.getKey().isPresent()) {
-      log.info("No new data, nothing to commit.. ");
-      return;
+      Optional<Dataset<Row>> transformed = dataAndCheckpoint.getLeft()
+              .map(data -> Optional.ofNullable(transformer.apply(jssc, data, props))).orElse(Optional.empty());
+      transformed.get().count();
+      checkpointStr = dataAndCheckpoint.getRight();
+      avroRDD = transformed.map(t ->
+         AvroConversionUtils.createRdd(t, "hudi_source", "hudi.source").toJavaRDD()
```

Comment:
Can we move this to CONSTANTS in the class ?

Reply:
Actually I see this being used below as well, may be needs to be a final static so can be shared.